Mathlib/Tactic/Algebraize.lean

```diff
@@ -0,0 +1,283 @@
+    -- Note: this could fail, but then `algebraize` will just continue, and won't add this instance
+    h.mvarId!.refl
+    let mvar ← mvarid.define nm tower
+      (← mkAppOptM ``IsScalarTower.of_algebraMap_eq'
+        #[l[0]!, l[1]!, l[2]!, none, none, none, none, none, none, h])
+    let (_, mvar) ← mvar.intro1P
+    return [mvar]
+
+/-- This function takes an array of expressions `t`, all of which are assumed to be `RingHom`s,
+and searches through the local context to find any additional properties of these `RingHoms`, after
+which it tries to add the corresponding `Algebra` properties to the context. It only looks for
+properties that have been tagged with the `algebraize` attribute, and uses this tag to find the
+corresponding `Algebra` property. -/
+def addProperties (t : Array Expr) : TacticM Unit := withMainContext do
+  let ctx ← getLCtx
+  ctx.forM fun decl => do
+    if decl.isImplementationDetail then return
+    let (nm, args) := decl.type.getAppFnArgs
+    -- Check if the type of the current hypothesis has been tagged with the `algebraize` attribute
+    match Attr.algebraizeAttr.getParam? (← getEnv) nm with
+    -- If it has, `p` will be the name of the corresponding `Algebra` property (or a constructor)
+    | some p =>
+      -- The last argument of the `RingHom` property is assumed to be `f`
+      let f := args[args.size - 1]!
+      -- Check that `f` appears in the list of functions given to `algebraize`
+      if ¬ (← t.anyM (Meta.isDefEq · f)) then return
+
+      let cinfo ← getConstInfo p
+      let n ← getExpectedNumArgs cinfo.type
+      let pargs := Array.mkArray n (none : Option Expr)
+      /- If the attribute points to the corresponding `Algebra` property itself, we assume that it
+      is definitionally the same as the `RingHom` property. Then, we just need to construct its type
+      and the local declaration will already give a valid term. -/
+      if cinfo.isInductive then
+        let pargs := pargs.set! 0 args[0]!
+        let pargs := pargs.set! 1 args[1]!
+        let tp ← mkAppOptM p pargs -- This should be the type `Algebra.Property A B`
+        liftMetaTactic fun mvarid => do
+          let nm ← mkFreshBinderNameForTactic `algebraizeInst
+          let mvar ← mvarid.define nm tp decl.toExpr
+          let (_, mvar) ← mvar.intro1P
+          return [mvar]
+      /- Otherwise, the attribute points to a constructor of the `Algebra` property. In this case,
+      we assume that the `RingHom` property is the last argument of the constructor (and that
+      this is all we need to supply explicitly). -/
+      else
+        let pargs := pargs.set! (n - 1) decl.toExpr
+        let val ← mkAppOptM p pargs
+        liftMetaTactic fun mvarid => do
+          let nm ← mkFreshBinderNameForTactic `algebraizeInst
+          -- let mvar ← mvarid.define nm _ val -- TODO: How to get the type of the target of `p`?
+          -- let (_, mvar) ← mvar.intro1P
+          let (_, mvar) ← mvarid.note nm val
```

Comment:
I have a hunch that some of the regression could come from the fact that we are using `MVarId.note` (`have` decl) instead of `MVarId.define` (`let` decl) here... Does anybody know how to construct the expected type of `val`?(added template as comment)

Reply:
I moved my debugging commits to

- leanprover-community/mathlib4#17284

Reply:
Very odd that it has regressed, given that an earlier bench showed no significant changes.

Reply:
Somehow I think the regression boils down to the import change (which shake suggested). Undoing that everything is fine again.